### ISSUE TITLE
Deoptionalization not detected

### ISSUE THREAD

**Issue by rkr-at-dbx** [python/mypy#2567]
Under `--strict-optional`, the following yields an error:

```
f=None  # type: Optional[Callable[..., None]]

# ...

if f is None:
    f = lambda *x: None
f()  # error: None not callable
```

Ideally, `f` should be understood not to be None at that point.

**Comment by Dakkaron** [comment]
I have a similar problem:

    input = "10" # this can be either "" or a string containing any long value
    value = long(input,0) if input!="" else None
    if value==None or value>=0x1000000000000000:
        pass
    if value==None or long(value)>=0x1000000000000000:
        pass
    if value==None or (value!=None and value>=0x1000000000000000):
        pass

I get the following errors:

    mypytestcase.py:3: error: Unsupported operand types for <= ("int" and "Optional[int]")
    mypytestcase.py:5: error: Argument 1 to "int" has incompatible type "Optional[int]"; expected "SupportsInt"
    mypytestcase.py:7: error: Unsupported operand types for <= ("int" and "Optional[int]")

I my opinion, all of these options should work.

**Comment by gvanrossum** [comment]
Probably, but have you tried replacing '== None' with 'is None' ?

**Comment by Dakkaron** [comment]
`is None` does fix the symptom, but I still thing that a None-Check with == should work.

**Comment by gvanrossum** [comment]
Well, a class could override `__eq__` to make its instances equal to `None`, so from `value == None` we cannot conclude that `value` is definitely `None`, and hence in the alternative branch (in your example, the `or value > ...` part) `value` *might* still be `None`.

If you want to continue discuss this please open a new issue; *this* issue should be about the original example (where assignment of a `lambda` doesn't seem to be interpreted right).

**Comment by ddfisher** [comment]
A simpler repro:
```py
f = None # type: Optional[Callable[[int], None]]
f = lambda x: None
f(0)  # error: None not callable
```

Inspecting things with `reveal_type`, it looks like the assignment on line 2 doesn't affect the binder for some reason.

**Comment by ddfisher** [comment]
Here's the root of the problem:
```py
reveal_type(lambda x: None)
```
```
test.py:1: error: Revealed type is 'Any'
```

Looking at `visit_func_expr` and `infer_lambda_type_using_context` in `checkexpr.py`, it appears that we infer `Any` for lambdas unless they have a `Callable` context.  We should be able to fix this by looking for a unique `Callable` in `Union` contexts.

**Comment by gvanrossum** [comment]
But why is a `lambda` not at least inferred as `Callable`? We know as much about it as we do about an unannotated `def`. This is really strange:
```python
def f(x): return x
g = lambda x: x
reveal_type(f)
reveal_type(g)
```
Gives
```
__tmp__.py:3: error: Revealed type is 'def (x: Any) -> Any'
__tmp__.py:4: error: Revealed type is 'Any'
```
I think they should reveal the same type.

**Comment by gvanrossum** [comment]
(Honestly this difference between `def` and `lambda` reminds me of a pretty common misconception about the difference between the two. E.g. many people mistakenly believe that only `lambda` supports closures. I guess we need a separate issue for that.)

**Comment by JukkaL** [comment]
Agreed that inferring `Any` as the type of a lambda should not happen silently.

**Comment by ddfisher** [comment]
+1.  This issue is fixed, but more work is needed to make lambda inference more robust.
